src/main/java/com/redhat/jenkins/plugins/cachet/pipeline/CachetGatingWaitForResourcesStep.java

```diff
@@ -0,0 +1,175 @@
+        return resources;
+    }
+
+    public int getTimeLimit() {
+        return timeLimit;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(timeLimit, resources, context);
+    }
+
+    public static class Execution extends SynchronousStepExecution<Map<String, Long>> {
+
+        private transient final int timeLimit;
+        private transient final List<String> resources;
+
+        private static final float RECURRENCE_PERIOD_BACKOFF = 1.2f;
+        static final long MIN_RECURRENCE_PERIOD = 15000; // ¼min
+        static final long MAX_RECURRENCE_PERIOD = 45000;
+        static long TIME_SO_FAR = 0;
+        long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+
+        Execution(int timeLimit, List<String> resources, StepContext context) {
+            super(context);
+            this.resources = resources;
+            this.timeLimit = timeLimit;
+        }
+
+        @Override
+        protected Map<String, Long> run() throws Exception {
+            Map<String, Long> waitForResourcesMet = new HashMap<>();
+            List<String> resourcesToWaitFor = new ArrayList<>();
+
+            Map<String, CachetGatingMetrics> metrics = getActionMetrics();
+            // First iteration over metrics to determine if there is a resource we should wait for
+            for (String resource : resources){
+                if(!metrics.containsKey(resource))
+                    throw new AbortException("One or more resources you want to wait for does not exist.");
+
+                CachetGatingMetrics tmpResourceMet = metrics.get(resource);
+                if(tmpResourceMet.getGatingStatus() == ResourceStatus.OPERATIONAL){
+                    waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                }
+                else {
+                    resourcesToWaitFor.add(resource);
+                }
+            }
+
+            long timeInMillis = TimeUnit.MINUTES.toMillis(timeLimit);
+            while (!resourcesToWaitFor.isEmpty()){// So there are resources we need to wait for
+                sleepingHandler(resourcesToWaitFor);
+
+                for (Iterator<String> iterator = resourcesToWaitFor.iterator(); iterator.hasNext();) {
+                    String resource = iterator.next();
+                    Resource tmpResource = getCurrentMetrics(resourcesToWaitFor).get(resource);
+                    if(tmpResource != null
+                            && tmpResource.getStatusName().equals("Operational")){
+                        iterator.remove();
+                        CachetGatingMetrics tmpResourceMet =  metrics.get(resource);
+                        tmpResourceMet.setGateUpdatedTime(System.currentTimeMillis());
+
+                        waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                    }
+                }
+                if(timeInMillis <= TIME_SO_FAR) break;
```

Comment:
Should the action you do here be an option?
- abort?
- break?

Reply:
I think it's a good idea!